src/__tests__/Project.test.js

```diff
@@ -390,4 +381,81 @@ describe('Project', () => {
 
     expect(ops).toEqual(['start:bar', 'end:bar', 'start:foo', 'end:foo']);
   });
+
+  test('runPackageTasks() fails with bail orderMode: serial', async done => {
+    let project = await Project.init(f.find('independent-workspaces'));
+    let packages = await project.getPackages();
+    let ops = [];
+
+    try {
+      await project.runPackageTasks(
+        packages,
+        { orderMode: 'serial', bail: true },
+        async pkg => {
+          ops.push('start:' + pkg.getName());
+          if (pkg.getName() === 'bar') {
+            throw new Error('test');
+          }
+          await Promise.resolve();
+          ops.push('end:' + pkg.getName());
+        }
+      );
+      done.fail(new Error('Error should have been thrown'));
+    } catch (error) {
+      expect(ops).toEqual(['start:bar']);
+    }
+    done();
+  });
+
+  test('runPackageTasks() fails with no bail orderMode: serial', async done => {
+    let project = await Project.init(f.find('independent-workspaces'));
+    let packages = await project.getPackages();
+    let ops = [];
+
+    try {
+      await project.runPackageTasks(
+        packages,
+        { orderMode: 'serial', bail: false },
+        async pkg => {
+          ops.push('start:' + pkg.getName());
+          if (pkg.getName() === 'bar') {
+            throw new Error('test');
+          }
+          await Promise.resolve();
+          ops.push('end:' + pkg.getName());
+        }
+      );
+      done.fail(new Error('Error should have been thrown'));
+    } catch (error) {
+      expect(ops).toEqual(['start:bar', 'start:foo', 'end:foo']);
+    }
+    done();
+  });
+
+  test('runPackageTasks() cleans after bail', async done => {
+    let project = await Project.init(f.find('independent-workspaces'));
+    let packages = await project.getPackages();
+    let ops = [];
+    let cleanupSpy = jest.fn(() => {});
+
+    try {
+      await project.runPackageTasks(
+        packages,
+        { orderMode: 'serial', bail: true },
+        async pkg => {
+          ops.push('start:' + pkg.getName());
+          if (pkg.getName() === 'bar') {
+            throw new Error('test');
+          }
+          await Promise.resolve();
+          ops.push('end:' + pkg.getName());
+        },
+        cleanupSpy
```

Comment:
I don't understand the point of cleanup?

Reply:
An issue I have seen is that the first task that throws an error (or returns a promise reject) causes the parent to also crash, leaving behind child process still running and logging to console.
Through the cleanup function, if the creator of task is creating many processes, the cleanup is to stop them if the bail flag is on

Reply:
Seems like this is what `try..finally` is for

Reply:
Just had a better idea :/ maybe this should be handled by `processes.js` directly with more `process.on` handlers...
This would remove the need for cleanup for the time being